chat-android/src/main/java/com/ably/chat/Messages.kt

```diff
@@ -170,25 +180,96 @@ data class SendMessageParams(
 }
 
-class DefaultMessages(
+internal class DefaultMessagesSubscription(
+    private val chatApi: ChatApi,
     private val roomId: String,
-    private val realtimeClient: RealtimeClient,
+    private val cancellation: Cancellation,
+    internal val fromSerialProvider: () -> DeferredValue<String>,
+) : MessagesSubscription {
+    override fun cancel() {
+        cancellation.cancel()
+    }
+
+    override suspend fun getPreviousMessages(start: Long?, end: Long?, limit: Int): PaginatedResult<Message> {
+        val fromSerial = fromSerialProvider().await()
+        val queryOptions = QueryOptions(start = start, end = end, limit = limit, orderBy = NewestFirst)
+        return chatApi.getMessages(
+            roomId = roomId,
+            options = queryOptions,
+            fromSerial = fromSerial,
+        )
+    }
+}
+
+internal class DefaultMessages(
+    private val roomId: String,
+    realtimeChannels: AblyRealtime.Channels,
     private val chatApi: ChatApi,
 ) : Messages {
 
+    private var listeners: Map<Messages.Listener, DeferredValue<String>> = emptyMap()
+
+    private var channelStateListener: ChannelStateListener
+
+    private var lock = Any()
+
     /**
      * the channel name for the chat messages channel.
      */
     private val messagesChannelName = "$roomId::\$chat::\$chatMessages"
 
-    override val channel: Channel
-        get() = realtimeClient.channels.get(messagesChannelName, ChatChannelOptions())
+    override val channel: Channel = realtimeChannels.get(messagesChannelName, ChatChannelOptions())
+
+    init {
+        channelStateListener = ChannelStateListener {
+            if (!it.resumed) updateChannelSerialsAfterDiscontinuity()
+        }
+        channel.on(ChannelState.attached, channelStateListener)
+    }
 
     override fun subscribe(listener: Messages.Listener): MessagesSubscription {
-        TODO("Not yet implemented")
+        val deferredChannelSerial = DeferredValue<String>()
+        addListener(listener, deferredChannelSerial)
+        val messageListener = PubSubMessageListener {
+            val pubSubMessage = it ?: throw AblyException.fromErrorInfo(
+                ErrorInfo("Got empty pubsub channel message", HttpStatusCodes.BadRequest, ErrorCodes.BadRequest),
+            )
+            val data = parsePubSubMessageData(pubSubMessage.data)
+            val chatMessage = Message(
+                roomId = roomId,
+                createdAt = pubSubMessage.timestamp,
+                clientId = pubSubMessage.clientId,
+                timeserial = pubSubMessage.extras.asJsonObject().requireString("timeserial"),
+                text = data.text,
+                metadata = data.metadata,
+                headers = pubSubMessage.extras.asJsonObject().get("headers")?.toMap() ?: mapOf(),
+            )
+            emitMessage(chatMessage)
+        }
+
+        channel.subscribe(MessageEventType.Created.eventName, messageListener)
```

Comment:
This would cause an implicit channel attach which will break room lifecycle - subscribing should cause no side-effects.

I assume there's an ably-java ticket that needs doing first? :)

Reply:
Yes, of course, https://github.com/ably/ably-java/issues/1027, I think until it release we can leave it as is. I'll add separate issue for this, too

Reply:
https://github.com/ably-labs/ably-chat-kotlin/issues/17